fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordPlannerTest.java

```diff
@@ -741,6 +751,298 @@ public void aggregateJoinIndex() throws Exception {
         }
     }
 
+    @Test
+    void joinOnNestedKeysWithDifferentTypes() throws Exception {
+        metaDataBuilder.getRecordType("CustomerWithHeader").setPrimaryKey(Key.Expressions.concat(field("___header").nest("z_key"), field("___header").nest("int_rec_id")));
+        metaDataBuilder.getRecordType("OrderWithHeader").setPrimaryKey(Key.Expressions.concat(field("___header").nest("z_key"), field("___header").nest("rec_id")));
+
```

Comment:
It might be more realistic to include something like
```
        metaDataBuilder.addIndex("OrderWithHeader", "order$custRef", field("custRef").nest("string_value"));
```
around here.

It does not affect correctness and the interesting case for this branch is the other direction, where a primary key scan's bounds come from a function inverse.